config/shared_config.go

```diff
@@ -210,78 +259,478 @@ func loadSharedConfig(ctx context.Context, configs configs) (Config, error) {
 //
 // For example, given two files A and B. Both define credentials. If the order
 // of the files are A then B, B's credential values will be used instead of A's.
-func NewSharedConfig(profile string, filenames []string) (SharedConfig, error) {
-	if len(filenames) == 0 {
-		return SharedConfig{}, fmt.Errorf("no shared config files provided")
+func NewSharedConfig(ctx context.Context, profile string, options sharedConfigsLoadInfo) (SharedConfig, error) {
+	if len(options.ConfigurationFiles) == 0 && len(options.CredentialsFiles) == 0 {
+		return SharedConfig{}, fmt.Errorf("no shared config or shared credentials options provided")
 	}
 
-	files, err := loadSharedConfigIniFiles(filenames)
+	// load shared configuration sections from shared configuration INI options
+	configSections, err := loadIniFiles(options.ConfigurationFiles)
 	if err != nil {
 		return SharedConfig{}, err
 	}
 
+	// check for profile prefix and drop duplicates or invalid profiles
+	err = processConfigSections(ctx, configSections, options.Logger)
+	if err != nil {
+		return SharedConfig{}, err
+	}
+
+	// load shared credentials sections from shared credentials INI options
+	credentialsSections, err := loadIniFiles(options.CredentialsFiles)
+	if err != nil {
+		return SharedConfig{}, err
+	}
+
+	// check for profile prefix and drop duplicates or invalid profiles
+	err = processCredentialsSections(ctx, credentialsSections, options.Logger)
+	if err != nil {
+		return SharedConfig{}, err
+	}
+
+	err = mergeSections(configSections, credentialsSections)
+	if err != nil {
+		return SharedConfig{}, err
+	}
+
+	// profile should be lower-cased to standardize
+	profile = strings.ToLower(profile)
+
 	cfg := SharedConfig{}
 	profiles := map[string]struct{}{}
-	if err = cfg.setFromIniFiles(profiles, profile, files); err != nil {
+	if err = cfg.setFromIniSections(profiles, profile, configSections, options.Logger); err != nil {
 		return SharedConfig{}, err
 	}
 
 	return cfg, nil
 }
 
-type sharedConfigFile struct {
+func processConfigSections(ctx context.Context, sections ini.Sections, logger logging.Logger) error {
+
+	prefix := "profile "
+	for _, section := range sections.List() {
+		// drop profiles without prefix for config files
+		if !strings.HasPrefix(section, prefix) && !strings.EqualFold(section, "default") {
+			// drop this section, as invalid profile name
+			sections.DeleteSection(section)
+
+			if logger != nil {
+				logger.Logf(logging.Debug,
+					"profile %v is ignored. A non-default profile without the \"profile \" prefix is invalid "+
+						"for the shared configuration file.\n",
+					section,
+				)
+			}
+		}
+	}
+
+	// rename sections to remove `profile ` prefixing to match with credentials file.
+	// if default is already present, it will be dropped.
+	for _, section := range sections.List() {
+		if strings.HasPrefix(section, prefix) {
+			v, ok := sections.GetSection(section)
+			if !ok {
```

Comment:
Also probably need a check that section name isn't empty after `profile ` prefix is stripped off.